### ISSUE TITLE
Ensure fonts appear in build directory

### ISSUE THREAD

**Issue by daikema** [CARTAvis/carta-archived#174]
This message had been showing up in the PureWeb logs when attempting to run the server version:

> HTTP 404 Not Found: /Apps/CartaVis-v2/html/common/skel/source/class/skel/theme/Roboto-Regular.ttf

This seems to be due to Qooxdoo being configured incorrectly as the error no longer appeared when reverting to an earlier version of [Font.js](https://github.com/CARTAvis/carta/blob/develop/carta/html5/common/skel/source/class/skel/theme/Font.js) which didn't specify a custom font.

According to the relevant bit of [the Qooxdoo documentation](http://www.qooxdoo.org/3.5.1/pages/desktop/ui_webfonts.html):

> Each source entry can be either a URI or a qooxdoo resource ID. The latter is generally preferable since font files will then be copied to the output directory for the build version just like any other application resource.

Basically the answer is to move to qooxdoo resource IDs for this (and any other similar file additions that I might not have noticed yet)

**Comment by daikema** [comment]
I'm wondering if some of the errors in the PureWeb logs that it was noted were also appearing on the Taiwan servers - i.e. things like `PageNotFound No mapping found for HTTP request` might also be due to a similar issue.

Are qooxdoo resource IDs being used anywhere?

**Comment by grimmerk** [comment]
@ajm-asiaa , @kswang1029 do you have time to take a look?

**Comment by ajm-ska** [comment]
OK. I'll check it out.

**Comment by ajm-ska** [comment]
We were getting several `PageNotFound No mapping found for HTTP request` errors here on our CARTA Pureweb server in Taiwan, although none related to the Roboto-Regular.ttf font. 
An example is below:
 
> 2017-07-19 10:23:52,545 INFO  18152 SessionController launch forwarding to /Apps/CartaSkeleton3/html5/server/serverIndex.html?app=CartaSkeleton3
> 2017-07-19 10:23:53,322 WARN        PageNotFound No mapping found for HTTP request with URI [/pureweb/app/unix:0.0/1/18152/server/pureweb.min.js] in DispatcherServlet with name 'cluster'
> 2017-07-19 10:23:53,323 ERROR       ErrorController An unexpected error occurred. Please contact your System Administrator with error identifier: 996 /pureweb/app/unix:0.0/1/18152/server/pureweb.min.js HTTP 404 Not Found from 140.109.177.16/140.109.177.16#543D3E78A93623EA2DD1640E507A8AD1
> 2017-07-19 10:23:53,423 WARN        PageNotFound No mapping found for HTTP request with URI [/pureweb/app/unix:0.0/1/18152/common/skel/source/script/skel.js] in DispatcherServlet with name 'cluster'
> 2017-07-19 10:23:53,423 WARN        PageNotFound No mapping found for HTTP request with URI [/pureweb/app/unix:0.0/1/18152/common/libs.js] in DispatcherServlet with name 'cluster'
> 2017-07-19 10:23:53,423 WARN        PageNotFound No mapping found for HTTP request with URI [/pureweb/app/unix:0.0/1/18152/server/serverConnector.js] in DispatcherServlet with name 'cluster'
> 2017-07-19 10:23:53,423 WARN        PageNotFound No mapping found for HTTP request with URI [/pureweb/app/unix:0.0/1/18152/common/CallbackList.js] in DispatcherServlet with name 'cluster'
> 2017-07-19 10:23:53,423 ERROR       ErrorController An unexpected error occurred. Please contact your System Administrator with error identifier: 997 /pureweb/app/unix:0.0/1/18152/common/skel/source/script/skel.js HTTP 404 Not Found from 140.109.177.16/140.109.177.16#543D3E78A93623EA2DD1640E507A8AD1
> 2017-07-19 10:23:53,423 ERROR       ErrorController An unexpected error occurred. Please contact your System Administrator with error identifier: 998 /pureweb/app/unix:0.0/1/18152/server/serverConnector.js HTTP 404 Not Found from 140.109.177.16/140.109.177.16#543D3E78A93623EA2DD1640E507A8AD1
> 2017-07-19 10:23:53,424 ERROR       ErrorController An unexpected error occurred. Please contact your System Administrator with error identifier: 1000 /pureweb/app/unix:0.0/1/18152/common/CallbackList.js HTTP 404 Not Found from 140.109.177.16/140.109.177.16#543D3E78A93623EA2DD1640E507A8AD1
> 2017-07-19 10:23:53,424 ERROR       ErrorController An unexpected error occurred. Please contact your System Administrator with error identifier: 999 /pureweb/app/unix:0.0/1/18152/common/libs.js HTTP 404 Not Found from 140.109.177.16/140.109.177.16#543D3E78A93623EA2DD1640E507A8AD1

Even though we were getting those errors, the CARTA server version still functioned fine.

But I found a way to not get those errors. 

Don't use the `getURLParameter(name)` function in `serverIndex.html` (comment out lines 11 to 19) and define the `base href` directly. e.g. In our case it is just `<base href="/Apps/CartaSkeleton3/html5/">` 

Please note: May need a different name depending on other setups. It is the name of your CARTA server version folder in the PureWeb Apps directory. For example our one is `/opt/CSI/PureWeb/Server/webapp/Apps/CartaSkeleton3`. I guess the idea of the `getURLParameter(name)` function was to allow it to work with any name.

Anyway, when that change is made, the CARTA server versions still works and we don't get any `PageNotFound No mapping found for HTTP request` errors in the log files.

Suggestion: Maybe we should unify what the CARTA folder should be named in the PureWeb Apps directory. e.g. `CARTAserver`. Then we can just use `<base href="/Apps/Cartaserver/html5/">` in `serverIndex.html`.

**Comment by daikema** [comment]
> We were getting several PageNotFound No mapping found for HTTP request errors here on our CARTA Pureweb server in Taiwan, although none related to the Roboto-Regular.ttf font.

Can you (at least temporarily) purge the font from all paths other than the build directory on your test machine and then retest?

I suspect it's likely already in some path that PureWeb picks up whereas the same isn't the case on our end.  Think you're using some CentOS release vs. Ubuntu 14.04 at our end so slightly different configs regarding things like fonts wouldn't surprise me.

> Please note: May need a different name depending on other setups. It is the name of your CARTA server version folder in the PureWeb Apps directory. For example our one is /opt/CSI/PureWeb/Server/webapp/Apps/CartaSkeleton3. I guess the idea of the getURLParameter(name) function was to allow it to work with any name.

Is there a PureWeb function that can be called to inject the name of the application (or perhaps have the javascript in question deduce this from its own path)?

**Comment by ajm-ska** [comment]
> Can you (at least temporarily) purge the font from all paths other than the build directory on your test machine and then retest?
> 
> I suspect it's likely already in some path that PureWeb picks up whereas the same isn't the case on our end. Think you're using some CentOS release vs. Ubuntu 14.04 at our end so slightly different configs regarding things like fonts wouldn't surprise me.

I checked our Docker server version, which is Ubuntu 14.04. The Roboto-Regular.ttf font was present at this location `/cartabuild/CARTAvis/carta/html5/common/skel/source/class/skel/theme/Roboto-Regular.ttf`. The only mention about it is the line `Debug: Not intercepting "file:///home/developer/src/CARTAvis/carta/VFS/DesktopDevel/common/skel/source/class/skel/theme/Roboto-Regular.ttf"` appearing in the terminal amongst other `Debug: Not intercepting`  lines related to qooxdoo. That is strange as the file actually does exist at that path. But either way, it is not the HTTP 404 error that you were receiving.

Then I checked an Ubuntu 16.04 virtual machine which I was setting up several months ago. There was no Roboto-Regular.ttf in the theme folder. I downloaded the latest CARTAvis from github, ran the `generate.py source-all` command and saw it was there. But then I noticed the timestamp on the font file was actually the same time that I downloaded the CARTAvis package. So I checked Github and found that Roboto-Regular.ttf is actually part of the package since 2 months ago; https://github.com/CARTAvis/carta/tree/develop/carta/html5/common/skel/source/class/skel/theme

Have you downloaded the latest develop version of CARTAvis recently?

> Is there a PureWeb function that can be called to inject the name of the application (or perhaps have the javascript in question deduce this from its own path)?

I suppose that is what the existing `getURIParameter(name)` function is doing. But I don't think it is really necessary. If we deploy the CARTA server version as a docker image, everything will already be set up. The same if we provide automated scripts to set up a server version on someone's machine. Plus, if we were to just provide setup instructions, we could explicitly state what to name it in the Pureweb Apps directory, or if the user wanted to call it something else, then remind them to also rename it in serverIndex.html.

**Comment by daikema** [comment]
> The Roboto-Regular.ttf font was present at this location /cartabuild/CARTAvis/carta/html5/common/skel/source/class/skel/theme/Roboto-Regular.ttf

Which is noticeably *NOT* in a PureWeb path - unless it's been duplicated there.  The doc snippet I mentioned earlier said that "font files will then be copied to the output directory" ... which if they're not (also) somewhere in the PureWeb directories doesn't appear to be the case.

> The only mention about it is the line Debug: Not intercepting "file:///home/developer/src/CARTAvis/carta/VFS/DesktopDevel/common/skel/source/class/skel/theme/Roboto-Regular.ttf" appearing in the terminal amongst other Debug: Not intercepting lines related to qooxdoo. That is strange as the file actually does exist at that path.

What I suspect that is being done in this case is that the font is being loaded directly through your computer rather than via tomcat/pureweb - i.e. `file:///etc/hosts` will probably load something in your browser if you're running OS X or Linux but you shouldn't be using such URIs when writing web pages.

To test, temporarily move or remove the CARTA directory once you think you've installed it in PureWeb (as files there shouldn't be needed anymore at that point) and then try running again.  (might only see an error on the debug console in your web browser - I'm not quite sure how `file://` URIs are treated).

**Comment by ajm-ska** [comment]
Maybe we are using Pureweb differently? In my `/opt/CSI/PureWeb/Server/webapp/Apps/CartaSkeleton3` directory, `html5` is a symbolic link to the the html5 directory in the CARTA build directory; `html5 -> /data/cartabuild/build_server/cpp/html5`. Similarly `server` is a symbolic link to `/data/cartabuild/build_server/cpp/server/server`.

I was able to reproduce your error by renaming the Roboto-Regular.ttf font to something else and re-running the `generate.py source-all` script. After starting up Pureweb and opening CARTA, I get the error in the log file:
```HTTP 404 Not Found: /Apps/CartaSkeleton3/html5/common/skel/source/class/skel/theme/Roboto-Regular.ttf```

When I renamed it back to the original Roboto-Regular.ttf and refreshed the browser page, it suddenly started working again (the font changed to Roboto-Regular). 

Then I renamed it to a wrong name again. But when I refresh the page, as well as stopping and starting the Pureweb server, the Roboto-Regular font is still appearing. So it must be somewhere in the browser cache now, as I think you suspect.

I can't go as far as moving entire CARTA directory. If I move the html5 directory and refresh the browser page, I get a `404 Not Found` in the browser.

Do you think we should not be defining a custom font in`/html5/common/skel/source/class/skel/theme/Font.js` e.g. Not defining Roboto-Regular.ttf, and just let the distribution and browser use whatever its default font is?

**Comment by daikema** [comment]
> Maybe we are using Pureweb differently? In my /opt/CSI/PureWeb/Server/webapp/Apps/CartaSkeleton3 directory, html5 is a symbolic link to the the html5 directory in the CARTA build directory; html5 -> /data/cartabuild/build_server/cpp/html5. Similarly server is a symbolic link to /data/cartabuild/build_server/cpp/server/server.

I'm trying to ensure that things are moved outside the build directory once the build is complete similar to how I'm setting up other software (i.e. generally removing the build directory at some point afterwards.  The command I'm using to copy get that dir populated is `./generate.py -m 'OPTIMIZE:[]' -m "BUILD_PATH:/opt/CSI/PureWeb/Server/webapp/Apps/CartaVis-v2/html" build`

**Comment by grimmerk** [comment]
@daikema I'm going to close it. It can be reopened if you have any concerns.
